app-backend/api/src/main/scala/shape/Routes.scala

```diff
@@ -44,12 +47,63 @@ trait ShapeRoutes extends Authentication
       get { listShapes } ~
       post { createShape }
     } ~
-    pathPrefix(JavaUUID) { shapeId =>
-      pathEndOrSingleSlash {
-        get { getShape(shapeId) } ~
-        put { updateShape(shapeId) } ~
-        delete { deleteShape(shapeId) }
+      post {
+        pathPrefix("upload") {
+          pathEndOrSingleSlash {
+            authenticate { user =>
+              val tempFile = ScalaFile.newTemporaryFile()
+              val response = storeUploadedFile("name", (_) => tempFile.toJava) { (m, _) =>
+                processShapefile(user, tempFile, m)
+              }
+              tempFile.delete()
+              response
+            }
+          }
+        } ~
+          pathPrefix(JavaUUID) { shapeId =>
+            pathEndOrSingleSlash {
+              get { getShape(shapeId) } ~
+                put { updateShape(shapeId) } ~
+                delete { deleteShape(shapeId) }
+            }
+          }
+      }
+  }
+
+  def processShapefile(user: User, tempFile: ScalaFile, fileMetadata: FileInfo) = {
+    val unzipped = tempFile.unzip()
+    val matches = unzipped.glob("*.shp")
+    matches.hasNext match {
+      case true => {
+        val shapeFile = matches.next()
+        val features = ShapeFileReader.readMultiPolygonFeatures(shapeFile.toString)
+
+        // Only reads first feature
```

Comment:
if we only read the first feature, we should probably let the user know in the final response.

Reply:
I'm not sure the best way to do that -- the response type is going to be a GeoJSON feature.